packages/grid/x-data-grid/src/tests/selection.DataGrid.test.tsx

```diff
@@ -571,6 +571,30 @@ describe('<DataGrid /> - Selection', () => {
       });
       expect(getSelectedRowIds()).to.deep.equal([0]);
     });
+
+    it('should retain the outdated selected rows when the rows prop changes when keepNonExistentRowsSelected is true', () => {
+      const data = getData(10, 2);
+      const onSelectionModelChange = spy();
+
+      const { setProps } = render(
+        <TestDataGridSelection
+          selectionModel={[0, 1, 2]}
+          checkboxSelection
+          keepNonExistentRowsSelected
+          onSelectionModelChange={onSelectionModelChange}
+          {...data}
+        />,
+      );
+      expect(getSelectedRowIds()).to.deep.equal([0, 1, 2]);
+
+      setProps({
+        rows: data.rows.slice(0, 1),
+      });
+      // This expectation is around visible rows
+      expect(getSelectedRowIds()).to.deep.equal([0]);
+      // The callback should be called with the true selectionModel
+      expect(onSelectionModelChange.lastCall.args[0]).to.deep.equal([0, 1, 2]);
```

Comment:
An alternative is to check the number of selected rows in the footer.

Reply:
This is partially why I added that call. I'll update the test and I should be able to remove all of it.